From: Ilya Dryomov <idryomov@gmail.com>
Date: Mon, 26 Sep 2016 15:43:52 +0200
Subject: rbd: add rbd_obj_request_error() helper
Git-commit: 0dcc685e7dd7190dcaa5435e9c14150f1d405b7b
Patch-mainline: v4.9-rc1
References: FATE#322288

Pull setting an error and marking a request done code into a new
helper.  obj_request_img_data_test() check isn't strictly needed right
now, but makes it applicable to !img_data requests and a bit safer.

Signed-off-by: Ilya Dryomov <idryomov@gmail.com>
Acked-by: Luis Henriques <lhenriques@suse.com>
---
 drivers/block/rbd.c |   28 ++++++++++++++++++----------
 1 file changed, 18 insertions(+), 10 deletions(-)

--- a/drivers/block/rbd.c
+++ b/drivers/block/rbd.c
@@ -1722,6 +1722,22 @@ static void rbd_osd_trivial_callback(str
 	obj_request_done_set(obj_request);
 }
 
+static void rbd_obj_request_error(struct rbd_obj_request *obj_request, int err)
+{
+	obj_request->result = err;
+	obj_request->xferred = 0;
+	/*
+	 * kludge - mirror rbd_obj_request_submit() to match a put in
+	 * rbd_img_obj_callback()
+	 */
+	if (obj_request_img_data_test(obj_request)) {
+		WARN_ON(obj_request->callback != rbd_img_obj_callback);
+		rbd_img_request_get(obj_request->img_request);
+	}
+	obj_request_done_set(obj_request);
+	rbd_obj_request_complete(obj_request);
+}
+
 static void rbd_osd_read_callback(struct rbd_obj_request *obj_request)
 {
 	struct rbd_img_request *img_request = NULL;
@@ -2859,11 +2875,7 @@ rbd_img_obj_parent_read_full_callback(st
 
 out_err:
 	ceph_release_page_vector(pages, page_count);
-	orig_request->result = img_result;
-	orig_request->xferred = 0;
-	rbd_img_request_get(orig_request->img_request);
-	obj_request_done_set(orig_request);
-	rbd_obj_request_complete(orig_request);
+	rbd_obj_request_error(orig_request, img_result);
 }
 
 /*
@@ -3014,11 +3026,7 @@ static void rbd_img_obj_exists_callback(
 	return;
 
 fail_orig_request:
-	orig_request->result = result;
-	orig_request->xferred = 0;
-	rbd_img_request_get(orig_request->img_request);
-	obj_request_done_set(orig_request);
-	rbd_obj_request_complete(orig_request);
+	rbd_obj_request_error(orig_request, result);
 }
 
 static int rbd_img_obj_exists_submit(struct rbd_obj_request *obj_request)
